bdk_core/src/chain_graph.rs

```diff
@@ -63,8 +63,37 @@ impl<I: ChainIndex> ChainGraph<I> {
         &self,
         update: &Self,
     ) -> Result<ChangeSet<I>, sparse_chain::UpdateFailure<I>> {
+        let mut chain_changeset = self.chain.determine_changeset(&update.chain)?;
+
+        // If we are invalidating checkpoints, it signals that a reorg has occured. We modify the
+        // changeset to add back "invalidated" transactions which do not conflict with transactions
+        // introduced by the `update`. We assume that all unconfirmed transactions in the `update`
+        // is included in the mempool.
+        if let Some(invalid_from) = chain_changeset.invalid_from {
+            // determine which original txs of the "invalidation range" should be added back
+            let retain_txs = self
+                .chain
+                .range_txids_by_height(TxHeight::Confirmed(invalid_from)..)
+                .filter_map(|&(_, txid)| self.graph.tx(txid).map(|tx| (txid, tx)))
+                // filter out txs that conflict with update's chain
+                .filter(|&(_, tx)| {
+                    update
+                        .graph
+                        .conflicting_txids(tx)
+                        .find(|(_, txid)| update.chain.tx_index(*txid).is_some())
+                        .is_none()
+                });
+
+            // add back non-conflicting txids to changeset
+            for (txid, _) in retain_txs {
+                if let Some(tx_index) = chain_changeset.txids.get_mut(&txid) {
+                    tx_index.get_or_insert(I::max_ord_of_height(TxHeight::Unconfirmed));
+                }
+            }
```

Comment:
Hmm I suppose this makes sense as a default policy. We probably want to provide ways of doing other policies eventually.